### DISCUSSION TITLE
Returning raw binary

### DISCUSSION THREAD

**Discussion by GeoSnipes** [nodejs/help#4860] [Q&A]
I have this function reference below. The issue is Whenever it returns data it seems to be double the size of the actual file. If I open the file in a file editor the contents look the same however when running fc /b and comparing the files nearly everything is different. How can I return the raw binary data?

 ```js
const fs = require('fs');
  
exports.handler = async (event) => {
	function getBinary (stream) {
		const chunks = [];
		let data = '';
		return new Promise((resolve) => {
			stream.on('data', (chunk) => {
				chunks.push(Buffer.from(chunk, "binary"));
				data += chunk;
			});
			stream.on('end', () => resolve(true ? Buffer.concat(chunks).toString('binary') : data)); 
		});
	}
	
	let readStream = fs.createReadStream('sourcepdf.pdf', {encoding: 'binary'});
	const result = await getBinary(readStream);
	return result;
};
```

**Comment by aduh95** [comment]
> How can I return the raw binary data?

One way of doing it would be to use [`readFile`](https://nodejs.org/api/fs.html#fs_fspromises_readfile_path_options):
```js
const fs = require('fs');
 
exports.handler = async (event) => {
   const result = await fs.promises.readFile('sourcepdf.pdf');
   return result;
};
```
That can even be a one-liner:
```js
const fs = require('fs/promises');
 
exports.handler = () => fs.readFile('sourcepdf.pdf');
```

Is there a reason for not using `readFile`?

> it seems to be double the size of the actual file

My guess would be it's because you convert the result to a string in your function, raw data is usually easier to deal with as `TypedArray` in JS (which is what `readFile` returns by default).
